### PR TITLE
Canonical Paths

### PR DESCRIPTION
**Pull request by Havvy**
Defines canonical paths, as part of #129.

Note to reviewer(s): There's also two commits that just change the location of links and change where newlines are, so just focus on the giant green blob in the paths file.

### CODE REVIEW

File: src/paths.md [hunk 1]
```diff
@@ -103,3 +92,79 @@ mod a {
...
+the crate.
+
+```rust
+// Comments show the canonical path of the item.
+
+mod a { // ::a
+    pub struct Struct; // ::a::Struct
+
+    pub trait Trait { // ::a::Trait
+        fn f(&self); // <::a::Trait>::f
```
Comment (petrochenkov) [inline src/paths.md]: `<::a::Trait>::f` -> `::a::Trait::f`.
`<::a::Trait>::f` has another meaning - inherent associated item on the trait object type `::a::Trait`.

Comment (Havvy) [reply src/paths.md]: I'm going with the assumption you are right, but I thought `::a::Trait::f` was syntactic sugar for `<::a::Trait>::f`.

File: src/paths.md [hunk 2]
```diff
@@ -103,3 +92,79 @@ mod a {
...
+
+## Canonical paths
+
+Items defined in a module or implementation have a *canonical path* that
+corresponds to where within its crate it is defined. All other paths to these
+items are aliases. The canonical path is defined as a *path prefix* appended by
+the path component the item itself defines.
+
+[Implementations] and [use declarations] do not have canonical paths, although
+the associated items that implementations define do have them. Items defined in
```
Comment (steveklabnik) [inline src/paths.md]: i wonder if "associated items" is too overloaded a term here, it made me do a double-take at first.

File: src/paths.md [hunk 3]
```diff
@@ -103,3 +92,79 @@ mod a {
...
+block expressions do not have canonical paths. Items defined in a module that
+does not have a canonical path do not have a canonical path. Associated items
+defined in an implementation that refers to an item without a canonical path,
+e.g. as the implementing type, the trait being implemented, a type parameter or
+bound on a type parameter, do not have canonical paths.
+
+The path prefix for modules is the canonical path to that module. For bare
+implementations, it is the canonical path of the item being implemented
+surrounded by angle (`<>`) brackets. For trait implementations, it is the
+cananonical path of the item being implemented followed by `as` followed by the
```
Comment (brauliobz) [inline src/paths.md]: cananonical -> canonical

Comment (Havvy) [reply src/paths.md]: Fixed.

File: src/paths.md (hunk 2 above)
Comment (Havvy) [reply src/paths.md]: It's technically extraneous, as implementations only define associated items. I removed it; does it flow better that way?

**Pull request merged**